File: apps/ai/src/models/interview.py
from pydantic import BaseModel


class DomainState(BaseModel):
    answered: int = 0
    total: int = 0
    vital_answered: bool = False


class InterviewState(BaseModel):
    project_type: str = "new"
    setup_confirmed: bool = False
    domains: dict[str, DomainState] = {}
    responses: dict[str, dict[str, str | bool]] = {}
    open_questions: list[str] = []
    documents_processed: list[str] = []
    conversation_summary: str = ""
    last_questions_asked: list[str] = []


class MessageEntry(BaseModel):
    role: str
    content: str


class ChatRequest(BaseModel):
    interview_id: str
    state: InterviewState
    history: list[MessageEntry] = []
    user_message: str
    documents_context: str | None = None
    tone_instruction: str | None = None
    is_resumption: bool = False
    llm_model: str = "gemini-2.5-flash"


class StateUpdate(BaseModel):
    state: InterviewState
    maturity: float


DOMAIN_WEIGHTS: dict[str, float] = {
    "scope": 0.30,
    "timeline": 0.20,
    "budget": 0.20,
    "integrations": 0.15,
    "tech_stack": 0.15,
}

VITAL_REQUIRED: dict[str, bool] = {
    "scope": True,
    "timeline": True,
    "budget": True,
    "integrations": False,
    "tech_stack": True,
}


def calculate_maturity(state: InterviewState) -> float:
    score = 0.0
    for domain, weight in DOMAIN_WEIGHTS.items():
        d = state.domains.get(domain)
        if not d:
            continue
        progress = min(d.answered / d.total, 1.0) if d.total > 0 else 0.0
        score += weight * progress

    all_vital = all(
        state.domains.get(domain, DomainState()).vital_answered
        for domain, required in VITAL_REQUIRED.items()
        if required
    )

    if not all_vital:
        score *= 0.85

    return round(score, 3)

File: apps/web/src/lib/types/interview.ts
export interface DomainState {
	answered: number;
	total: number;
	vital_answered: boolean;
}

export interface InterviewState {
	project_type: string;
	setup_confirmed: boolean;
	domains: Record<string, DomainState>;
	responses: Record<string, { value: string; source: string; confirmed: boolean }>;
	open_questions: string[];
	documents_processed: string[];
	conversation_summary: string;
	last_questions_asked: string[];
}

export interface Interview {
	id: string;
	user_id: string;
	title: string | null;
	status: string;
	state: InterviewState;
	maturity: number;
	created_at: Date;
	updated_at: Date;
}

export interface InterviewMessage {
	id: string;
	interview_id: string;
	role: string;
	content: string;
	tokens_used: number;
	created_at: Date;
}

export interface InterviewDocument {
	id: string;
	interview_id: string;
	filename: string;
	mime_type: string;
	storage_path: string;
	file_hash: string | null;
	extracted_text: string | null;
	status: string;
	created_at: Date;
}

export const DOMAIN_WEIGHTS: Record<string, number> = {
	scope: 0.3,
	timeline: 0.2,
	budget: 0.2,
	integrations: 0.15,
	tech_stack: 0.15,
};

export const VITAL_REQUIRED: Record<string, boolean> = {
	scope: true,
	timeline: true,
	budget: true,
	integrations: false,
	tech_stack: true,
};

export const MATURITY_THRESHOLD = 0.7;

export function createDefaultState(): InterviewState {
	return {
		project_type: 'new',
		setup_confirmed: false,
		domains: {
			scope: { answered: 0, total: 8, vital_answered: false },
			timeline: { answered: 0, total: 5, vital_answered: false },
			budget: { answered: 0, total: 4, vital_answered: false },
			integrations: { answered: 0, total: 6, vital_answered: false },
			tech_stack: { answered: 0, total: 5, vital_answered: false },
		},
		responses: {},
		open_questions: [],
		documents_processed: [],
		conversation_summary: '',
		last_questions_asked: [],
	};
}

export function calculateMaturity(state: InterviewState): number {
	const { domains } = state;
	let score = 0;

	for (const [domain, weight] of Object.entries(DOMAIN_WEIGHTS)) {
		const d = domains[domain];
		if (!d) continue;
		const progress = d.total > 0 ? Math.min(d.answered / d.total, 1.0) : 0;
		score += weight * progress;
	}

	const allVitalAnswered = Object.entries(VITAL_REQUIRED).every(([domain, required]) => {
		if (!required) return true;
		return domains[domain]?.vital_answered ?? false;
	});

	if (!allVitalAnswered) {
		score *= 0.85;
	}

	return Math.round(score * 1000) / 1000;
}
